File: packages/kiwicaptcha-risk-php/tools/fixture_hash.php
<?php

declare(strict_types=1);

/**
 * Prints the sha256 of the cross-language risk-v1 protocol parity blob.
 *
 * The blob order is fixed and mirrored byte-for-byte by the Rust side
 * (`packages/kiwicaptcha-risk/examples/fixture_hash.rs`); CI's
 * risk-parity job compares the two hashes:
 *
 *   1. every fixtures.json case's big-endian u16 contract score;
 *   2. the five hkdf-sha256 identity keys of identity_vectors.master_key
 *      in source, subnet, session, principal, event order (32 raw bytes
 *      each);
 *   3. canonical IP bytes and masked networks for the fixed probe list
 *      (IPv4, IPv6, IPv4-mapped IPv6 normalized to IPv4);
 *   4. the identity_vectors derivations as raw 16-byte pseudonyms:
 *      session (from the decoded cookie bytes), principal, source and
 *      subnet (at the fixture epochs);
 *   5. every namespace_vectors derived namespace (utf-8 bytes).
 *
 * Steps 2-5 make this a real protocol parity hash: HKDF keys, canonical
 * IP encoding, IPv4-mapped normalization, subnet masking, the ephemeral
 * pseudonyms and the deployment namespace are all compared across the
 * two implementations, not only the score vector.
 */

require dirname(__DIR__) . '/vendor/autoload.php';

use KiwiCaptcha\Risk\DeploymentNamespace;
use KiwiCaptcha\Risk\RiskIdentityFactory;
use KiwiCaptcha\Risk\RiskKeys;
use KiwiCaptcha\Risk\RiskScorer;
use KiwiCaptcha\Risk\RiskWeights;
use KiwiCaptcha\Risk\SignalVector;

/** The canonical probe list, mirrored by the Rust tool. */
const PROBE_IPS = [
    '203.0.113.27',
    '192.0.2.1',
    '2001:db8:abcd:12ff:ffff:ffff:ffff:ffff',
    '::ffff:203.0.113.27',
];

$path = dirname(__DIR__) . '/../../protocol/risk-v1/fixtures.json';
$raw = @file_get_contents($path);
if ($raw === false) {
    fwrite(STDERR, sprintf("fixtures.json not found at %s\n", $path));
    exit(1);
}
$fixtures = json_decode($raw, true);
if (!is_array($fixtures) || ($fixtures['protocol'] ?? null) !== 'risk-v1') {
    fwrite(STDERR, "fixtures.json must be a risk-v1 document\n");
    exit(1);
}

$blob = '';

// 1. Contract scores over the shared signal fixtures.
$weights = RiskWeights::fromArray($fixtures['weights']);
$scorer = new RiskScorer();
$base = (int) $fixtures['base_risk'];
foreach ($fixtures['fixtures'] as $fixture) {
    $blob .= pack('n', $scorer->score($base, SignalVector::fromArray($fixture['signals']), $weights));
}

// 2. HKDF identity keys from the shared master.
$identity = $fixtures['identity_vectors'];
$keys = RiskKeys::fromMaster($identity['master_key']);
$blob .= $keys->source . $keys->subnet . $keys->session . $keys->principal . $keys->event;

// 3. Canonical IP encodings and masked networks.
$factory = new RiskIdentityFactory($keys);
foreach (PROBE_IPS as $probe) {
    $blob .= $factory->canonicalIp($probe) . $factory->maskIp($probe);
}

// 4. The identity golden vectors as raw pseudonym bytes.
$blob .= (string) hex2bin($factory->sessionId($identity['session']['cookie_hex']));
$blob .= (string) hex2bin($factory->principalId($identity['principal']['material_utf8']));
$blob .= (string) hex2bin($factory->sourceId($identity['source']['ip'], (int) $identity['source']['epoch'] * 900));
$blob .= (string) hex2bin($factory->subnetId($identity['subnet']['ip'], (int) $identity['subnet']['epoch'] * 900));

// 5. Deployment namespace derivations from the shared vectors.
foreach ($fixtures['namespace_vectors'] as $vector) {
    $blob .= DeploymentNamespace::derive($vector['raw'], (int) $vector['version']);
}

echo hash('sha256', $blob), "\n";

File: packages/kiwicaptcha-risk/examples/fixture_hash.rs
//! Prints the sha256 of the cross-language risk-v1 protocol parity blob.
//!
//! The blob order is fixed and mirrored byte-for-byte by the PHP side
//! (`packages/kiwicaptcha-risk-php/tools/fixture_hash.php`); CI's
//! risk-parity job compares the two hashes:
//!
//!   1. every fixtures.json case's big-endian u16 contract score;
//!   2. the five hkdf-sha256 identity keys of identity_vectors.master_key
//!      in source, subnet, session, principal, event order (32 raw bytes
//!      each);
//!   3. canonical IP bytes and masked networks for the fixed probe list
//!      (IPv4, IPv6, IPv4-mapped IPv6 normalized to IPv4);
//!   4. the identity_vectors derivations as raw 16-byte pseudonyms:
//!      session (from the decoded cookie bytes), principal, source and
//!      subnet (at the fixture epochs);
//!   5. every namespace_vectors derived namespace (utf-8 bytes).
//!
//! Steps 2-5 make this a real protocol parity hash: HKDF keys, canonical
//! IP encoding, IPv4-mapped normalization, subnet masking, the ephemeral
//! pseudonyms and the deployment namespace are all compared across the
//! two implementations, not only the score vector.

use kiwicaptcha_risk::identity::{canonical_ip, masked_network, RiskIdentityFactory};
use kiwicaptcha_risk::keys::RiskKeys;
use kiwicaptcha_risk::namespace::{deployment_namespace, NamespaceVersion};
use kiwicaptcha_risk::score::{score, RiskWeights};
use kiwicaptcha_risk::signals::SignalVector;
use sha2::{Digest, Sha256};
use std::net::IpAddr;

/// The canonical probe list: IPv4, IPv6 and an IPv4-mapped IPv6 that must
/// normalize to its 4-byte IPv4 form on both sides.
const PROBE_IPS: [&str; 4] = [
    "203.0.113.27",
    "192.0.2.1",
    "2001:db8:abcd:12ff:ffff:ffff:ffff:ffff",
    "::ffff:203.0.113.27",
];

fn main() {
    let path = concat!(
        env!("CARGO_MANIFEST_DIR"),
        "/../../protocol/risk-v1/fixtures.json"
    );
    let raw = std::fs::read_to_string(path).expect("fixtures.json must exist at the repo root");
    let fixtures: serde_json::Value = serde_json::from_str(&raw).expect("fixtures.json must parse");
    assert_eq!(fixtures["protocol"], "risk-v1");

    let mut blob: Vec<u8> = Vec::new();

    // 1. Contract scores over the shared signal fixtures.
    let weights: RiskWeights =
        serde_json::from_value(fixtures["weights"].clone()).expect("weights decode");
    let base = fixtures["base_risk"].as_u64().expect("base_risk") as u16;
    for case in fixtures["fixtures"].as_array().expect("fixtures array") {
        let vector: SignalVector =
            serde_json::from_value(case["signals"].clone()).expect("signals decode");
        blob.extend_from_slice(&score(base, &vector, &weights).to_be_bytes());
    }

    // 2. HKDF identity keys from the shared master.
    let identity = &fixtures["identity_vectors"];
    let master = identity["master_key"].as_str().expect("master_key");
    let keys = RiskKeys::from_master(master.as_bytes());
    for key in [
        &keys.source,
        &keys.subnet,
        &keys.session,
        &keys.principal,
        &keys.event,
    ] {
        blob.extend_from_slice(key);
    }

    // 3. Canonical IP encodings and masked networks.
    let factory = RiskIdentityFactory::new(keys);
    for probe in PROBE_IPS {
        let ip: IpAddr = probe.parse().expect("probe IP parses");
        blob.extend_from_slice(&canonical_ip(ip));
        blob.extend_from_slice(&masked_network(ip, 24, 56));
    }

    // 4. The identity golden vectors as raw pseudonym bytes.
    let cookie = identity["session"]["cookie_hex"]
        .as_str()
        .expect("cookie_hex");
    blob.extend_from_slice(&factory.session_id(&hex::decode(cookie).expect("cookie hex")));
    let principal = identity["principal"]["material_utf8"]
        .as_str()
        .expect("material_utf8");
    blob.extend_from_slice(&factory.principal_id(principal.as_bytes()));
    let source_ip: IpAddr = identity["source"]["ip"].as_str().unwrap().parse().unwrap();
    let source_epoch = identity["source"]["epoch"].as_i64().unwrap();
    blob.extend_from_slice(
        hex::decode(factory.source_id(source_ip, source_epoch * 900))
            .unwrap()
            .as_slice(),
    );
    let subnet_ip: IpAddr = identity["subnet"]["ip"].as_str().unwrap().parse().unwrap();
    let subnet_epoch = identity["subnet"]["epoch"].as_i64().unwrap();
    blob.extend_from_slice(
        hex::decode(factory.subnet_id(subnet_ip, subnet_epoch * 900))
            .unwrap()
            .as_slice(),
    );

    // 5. Deployment namespace derivations from the shared vectors.
    for vector in fixtures["namespace_vectors"]
        .as_array()
        .expect("namespace_vectors")
    {
        let raw_ns = vector["raw"].as_str().expect("raw namespace");
        let version = NamespaceVersion::from_u8(vector["version"].as_u64().unwrap() as u8)
            .expect("known namespace version");
        blob.extend_from_slice(deployment_namespace(raw_ns, version).as_bytes());
    }

    println!("{}", hex::encode(Sha256::digest(&blob)));
}
